### DISCUSSION TITLE
Possibilities to get typehints for OPCUA structures in editor?

### DISCUSSION THREAD

**Discussion by rempAut** [FreeOpcUa/opcua-asyncio#1752] [General]
Hello, I am writing a opcua-asyncio server application which communicates with a PLC. 

We have a few structures defined for our communication. What I am trying to accomplish is getting typehints in my editor for these structures. So far I've tried:

### 1. Creating dataclasses, then manually creating OPCUA structures with `new_struct` from these dataclasses on the server
this seemingly solves the problem, There are typehints when I edit the code, but there are 2 places where I have to manually maintain the structures (in the dataclasses and in the actual `new_struct` calls). I don't like it.

### 2. Importing the OPCUA structures from a nodeset file with `import_xml`
The idea behind this was to autogenerate dataclasses code from a nodeset file. We would still need to import the correct latest nodeset, but at least the nodeset file becomes the single maintenance point for both the server and client. 
The structures are imported in the `ua` module and I can inspect the generated dataclasses, but there are no typehints when I edit the code, and I have no idea how could I get them.

I'd like to hear other people's ideas and opinions about this :-)

### pic: that's what I'm looking for
![image](https://github.com/user-attachments/assets/2c0a1449-7519-4d29-bd23-2c80d56876d7)

**Comment by schroeder-** [comment]
You can create script, that parses the nodeset xml and then create a python file like this for typ hints:

```
from typing import TYPE_CHECKING
import ua
if TYPE_CHECKING:
     @dataclass
      class plcParentStructType:
          parent_id: ua.Int32
          name: ua.String
          children: list[plcChildStructType]
```

or you could make a test server with a modified version of asyncua, which then loads the nodes and outputs the conten of:
https://github.com/FreeOpcUa/opcua-asyncio/blob/302c434379490aba8925b9d925de26581a482633/asyncua/common/structures104.py#L306 into a file.
